src/search.jl

```diff
@@ -544,3 +546,410 @@ function execute(search::GeneticAlgorithmSearch, problem)
         end
     end
 end
+
+
+#=
+Julia - GSOC
+
+Ahmed Madbouly
+
+Implementation for AND-OR, Online DFS, LRTA
+=#
+
+abstract type AbstractProblemStruct end;
+
+mutable struct ProblemStruct <: AbstractProblemStruct
+    initial::String
+    goal::Nullable{String}
+
+    function ProblemStruct(initial_state::String; goal_state::Union{Void, String}=nothing)
+        return new(initial_state, Nullable{String}(goal_state));
+    end
+
+end
+
+#=
+    OnlineDFSAgentProgram is a online depth first search agent (Fig. 4.21)
+=#
+mutable struct OnlineDFSAgentProgram <: AgentProgram
+    result::Dict
+    untried::Dict
+    unbacktracked::Dict
+    state::Nullable{String}
+    action::Nullable{String}
+    problem::AbstractProblemStruct
+
+    function OnlineDFSAgentProgram{T <: AbstractProblemStruct}(problem::T)
+        return new(Dict(), Dict(), Dict(), Nullable{String}(), Nullable{String}(), problem);
+    end
+end
+
+struct MemoizedFunction
+    f::Function     #original function
+    values::Dict{Tuple{Vararg}, Any}
+
```

Comment:
Vararg and Any for a Dict looks like we have no understanding of the underlying system. Julia has a very structured type definition with templates can you see what can be reasonably be used there?